packages/versioning/src/lib.rs

```diff
@@ -94,33 +94,45 @@ pub fn initialize(storage: &mut dyn Storage, version: Version) -> StdResult<()>
     VERSION_STORAGE_KEY.save(storage, &version)
 }
 
-pub fn update_software(storage: &mut dyn Storage, new: Version) -> Result<ReleaseLabel, StdError> {
+pub fn update_software<ContractError, MapErrorFunctor>(
+    storage: &mut dyn Storage,
+    new: Version,
+    map_error: MapErrorFunctor,
+) -> Result<ReleaseLabel, ContractError>
+where
+    MapErrorFunctor: Fn(StdError) -> ContractError,
+{
     load_version(storage)
-        .and_then(|current| release::allow_software_update(&current, &new))
-        .and_then(|()| save_version(storage, &new))
+        .map_err(&map_error)
+        .and_then(|current| release::allow_software_update(&current, &new).map_err(&map_error))
+        .and_then(|()| save_version(storage, &new).map_err(map_error))
         .map(|()| release::label())
```

Comment:
is not possible to `.map_err` only once at the end?

Reply:
That's a very good point. Thank you!